src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp

```diff
@@ -1196,121 +1144,131 @@ Array<Pair<Index, Pass>> Orbit::ComputePasses(const Array<State>& aStateArray, c
-            }
+        std::tie(northPointCrossing, descendingNodeCrossing, southPointCrossing, passBreakCrossing) =
+            Orbit::ComputeCrossings(aModel, previousInstant, stepDuration, isForwardPropagated, anEndInstant);
 
-            // Pass break
-            if ((previousPositionCoordinates_ECI.z() < 0.0) && (currentPositionCoordinates_ECI.z() >= 0.0))
-            {
-                passBreak =
-                    Orbit::GetCrossingInstant(epoch, previousStatePtr->accessInstant(), state.accessInstant(), getZ);
-            }
+        Pass pass = {
+            revolutionNumber,
+            isForwardPropagated ? previousPassEndInstant : passBreakCrossing,
+            northPointCrossing,
+            descendingNodeCrossing,
+            southPointCrossing,
+            isForwardPropagated ? passBreakCrossing : previousPassEndInstant,
+        };
 
-            if (passBreak.isDefined())
-            {
-                const Pass pass = {
-                    revolutionNumber,
-                    previousPassEndInstant,
-                    northPointCrossing,
-                    descendingNodeCrossing,
-                    southPointCrossing,
-                    passBreak,
-                };
+        passes.add(pass);
 
-                passMap.add({currentIndex, pass});
+        // if forward propagating and the pass has an undefined PassBreak, then we have crossed the end of the interval
+        // if backward propagating and the pass has an undefined AscendingNode, then we have crossed the end of the
+        // interval
+        if ((isForwardPropagated && !pass.accessInstantAtPassBreak().isDefined()) ||
+            (!isForwardPropagated && !pass.accessInstantAtAscendingNode().isDefined()))
+        {
+            break;
+        }
 
-                revolutionNumber++;
-                previousPassEndInstant = passBreak;
+        revolutionNumber += isForwardPropagated ? 1 : -1;
+        previousPassEndInstant = passBreakCrossing;
 
-                southPointCrossing = Instant::Undefined();
-                northPointCrossing = Instant::Undefined();
-                descendingNodeCrossing = Instant::Undefined();
-                passBreak = Instant::Undefined();
-            }
+        if (pass.isComplete())
+        {
+            Array<Duration> durations = {
+                (pass.accessInstantAtNorthPoint() - pass.accessInstantAtAscendingNode()),
+                (pass.accessInstantAtDescendingNode() - pass.accessInstantAtNorthPoint()),
+                (pass.accessInstantAtSouthPoint() - pass.accessInstantAtDescendingNode()),
+                (pass.accessInstantAtPassBreak() - pass.accessInstantAtSouthPoint()),
+            };
+            stepDuration = *std::min_element(durations.begin(), durations.end()) / 2.0;
         }
 
-        previousStatePtr = &state;
-        currentIndex++;
+        previousInstant = isForwardPropagated ? pass.accessInstantAtPassBreak() + Duration::Microseconds(1.0)
```

Comment:
Microsecond is a good enough time interval? It doesn't work with something like a nanosecond? or 10 nanoseconds?

Reply:
Yeah unless we have passes that are less than 1 microsecond long. Which I can live with :P
This is more than sufficient, we could even do 1 second or 100 seconds and be fine.